fish-rust/src/fd_monitor.rs

```diff
@@ -0,0 +1,573 @@
+
+            // A predicate which services each item in turn, returning true if it should be removed
+            let servicer = |item: &mut FdMonitorItem| {
+                let fd = item.fd.as_raw_fd();
+                if item.service_item(&fds, &now) == ItemAction::Remove {
+                    FLOG!(fd_monitor, "Removing fd", fd);
+                    return ItemAction::Remove;
+                }
+                return ItemAction::Retain;
+            };
+
+            // Service all items that are either readable or have timed out, and remove any which
+            // say to do so.
+
+            // This line is from the C++ codebase (fd_monitor.cpp:170) but this write is never read.
+            // now = Instant::now();
+
+            self.items
+                .retain_mut(|item| servicer(item) == ItemAction::Retain);
+
+            // Handle any changes if the change signaller was set. Alternatively, this may be the
+            // wait lap, in which case we might want to commit to exiting.
+            let change_signalled = fds.test(change_signal_fd);
+            if change_signalled || is_wait_lap {
+                // Clear the change signaller before processing incoming changes
+                self.change_signaller.try_consume();
+                let mut data = self.data.lock().expect("Mutex poisoned!");
+
+                // Move from `pending` to the end of `items`
+                self.items.extend(&mut data.pending.drain(..));
+
+                // Grab any poke list
+                assert!(
+                    pokelist.is_empty(),
+                    "poke list should be empty or else we're dropping pokes!"
+                );
+                std::mem::swap(&mut pokelist, &mut data.pokelist);
+
+                if data.terminate
+                    || (is_wait_lap
+                        && self.items.is_empty()
+                        && pokelist.is_empty()
+                        && !change_signalled)
+                {
+                    // Maybe terminate is set. Alternatively, maybe we had no items, waited a bit,
+                    // and still have no items. It's important to do this while holding the lock,
+                    // otherwise we race with new items being added.
+                    assert!(
+                        data.running,
+                        "Thread should be running because we're that thread"
+                    );
+                    FLOG!(fd_monitor, "Thread exiting");
+                    data.running = false;
+                    break;
+                }
+            }
+        }
+    }
+
+    /// Poke items in the poke list, removing any items that close their fd in their callback. The
+    /// poke list is consumed after this. This is only called from the background thread.
+    fn poke(&mut self, pokelist: &[FdMonitorItemId]) {
+        self.items.retain_mut(|item| {
+            let fd = item.fd.as_raw_fd();
+            let action = item.maybe_poke_item(&*pokelist);
```

Comment:
It's not. I got into the habit because that syntax is sometimes needed (it's called an "explicit reborrow") such as if you need to pass a mut reference to `self` to another function in a function that's already using using `&mut self`. 

Personal preference as well to always be clear about borrows vs moves which isn't always obvious just by looking at the call site. Clippy doesn't like the unneeded ref + deref either by default, I guess we'll have to decide on what we want to unify around so it's all the same.